### PR TITLE
adds linkdata component

### LINKED ISSUE #577: Add Schema.org metadata
**Is your feature request related to a problem? Please describe.**
Our article page is currently missing metadata. This PR will add all the available Schema.org meta.

NB, the same solution [can be applied to AMP as it allows JSON LD scripts](https://www.ampproject.org/docs/fundamentals/discovery#use-schema.org-for-most-search-engines).

**Describe the solution you'd like**
Add the following fields, including the logic necessary to render them correctly:
- `url` - canonical URL
- `datePublished` - `metadata.firstPublished` - convert to ISO 8601 format (see [docs](https://schema.org/Date))
- `dateModified` - `metadata.lastPublished` - convert to ISO 8601 format (see [docs](https://schema.org/Date))
- `headline` - `promo.headlines.seoHeadline`

The final structure should look like this (from Jim's comment below):
```
<script type="application/ld+json">
    {
      "@context": "http://schema.org",
      "@type": "ReportageNewsArticle",
      "url": "https://www.bbc.com/news/articles/[OPTIMO ID]",
      "publisher": {
        "@type": "NewsMediaOrganization",
        "name": "BBC News",
        "publishingPrinciples": "http://www.bbc.com/news/help-41670342",
        "logo": {
          "@type": "ImageObject",
          "width": 1024,
          "height": 576,
          "url": "https://www.bbc.com/news/special/2015/newsspec_10857/bbc_news_logo.png?cb=1"
        }
      },
      "datePublished": "[FIRSTPUB e.g. 2018-09-05T14:35:15+01:00]",
      "dateModified": "[LASTPUB e.g. 2018-09-05T14:35:15+01:00]",
      "headline": "[SEO HEADLINE]",
      "image": {
        "@type": "ImageObject",
        "width": 1024,
        "height": 576,
        "url": "https://www.bbc.com/news/special/2015/newsspec_10857/bbc_news_logo.png?cb=1"
      },
      "thumbnailUrl": "https://www.bbc.com/news/special/2015/newsspec_10857/bbc_news_logo.png?cb=1",
      "author": {
        "@type": "NewsMediaOrganization",
        "name": "BBC News",
        "logo": {
          "@type": "ImageObject",
          "width": 1024,
          "height": 576,
          "url": "https://www.bbc.com/news/special/2015/newsspec_10857/bbc_news_logo.png?cb=1"
        },
        "noBylinesPolicy": "http://www.bbc.com/news/help-41670342#authorexpertise"
      },
      
    }

    </script>
```

**Additional context**
Some fields might require further discussion, and we will initially hard-code them.

**Testing notes**
- Add Cypress tests, which will duplicate (and then amend) current logic found under `retrieveMetaDataContent` but aiming at looking under the `head` for the metadata.
- Unit tests are adequate

| Optimo Category | Type |
| --------------- | ---- |
| Analysis | AnalysisNewsArticle |
| Ask The Audience | AskPublicNewsArticle |
| Explainer | BackgroundNewsArticle |
| Opinion | OpinionNewsArticle |
| News or Feature | ReportageNewsArticle |
| Review | ReportageNewsArticle |
| Fact check | ReportageNewsArticle |
| Summary | ReportageNewsArticle |
| Polls and Surveys | ReportageNewsArticle |

### PR DESCRIPTION
**Pull request by twinlensreflex**
Resolves #577 

**Overall change:** _Adds LinkData component for metadata._

**Code changes:**

- A new LinkData component has been added

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] Tests added for new features
- [x] Test engineer approval
- [x] I have followed [the merging checklist](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/MERGE_PROCESS.md) and this is ready to merge.

### CODE REVIEW

File: src/app/components/LinkData/index.jsx [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  lastUpdated,
+  optimoUrn,
+  service,
+}) => {
+  const imgObject = 'ImageObject';
+  const newsMediaOrg = 'NewsMediaOrganization';
+  const url =
+    'https://www.bbc.com/news/special/2015/newsspec_10857/bbc_news_logo.png?cb=1';
+
+  const optimoId = optimoUrn.split(':').pop();
```
Comment (sareh) [inline src/app/components/LinkData/index.jsx]: To remove duplication of this logic between the Metadata Container and here, you could use LinkData component in the container here: https://github.com/bbc/simorgh/blob/latest/src/app/containers/Metadata/index.jsx#L69
So passing the values `optimoId`, `timeFirstPublished`, `timeLastPublished` into this component.

Comment (twinlensreflex) [reply src/app/components/LinkData/index.jsx]: great call, thanks :D I made the changes, but the <script> tag still doesn't show in the head.. I tried copying exactly how the `Metadata` component uses Helmet (ie. with the htmlAttributes) but it didn't work. 

https://github.com/bbc/simorgh/blob/a6a08046c37b60d231c981c97e1a99407f32162c/src/server/Document/component.jsx#L14 this is coming through as an empty array... not sure why 😬 and not really sure how to fix it...

**Comment by sareh** [comment]
There is a Cypress integration test in `./cypress/integration/ampSpec.js` that is checking that the first `<script>` tag in the head is the AMP script. This is no longer true, so this test is failing. 
You should be able to update the integration test so that it passes.

File: src/app/components/LinkData/__snapshots__/index.test.jsx.snap [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`LinkData should correctly render metadata for links 1`] = `null`;
```
Comment (dr3) [inline src/app/components/LinkData/__snapshots__/index.test.jsx.snap]: This snapshot isnt snapshotting anything (it has problems with helmet), you might need to pick a different snapshot render depth

File: src/app/components/LinkData/index.jsx [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    '@type': imgObject,
+    width: 1024,
+    height: 576,
+    url,
+  };
+
+  const publisher = {
+    '@type': newsMediaOrg,
+    name: service,
+    publishingPrinciples: 'http://www.bbc.com/news/help-41670342',
```
Comment (dr3) [inline src/app/components/LinkData/index.jsx]: Can this data come from the news context?

**Review by dr3 — CHANGES_REQUESTED** [review]
Looks good, just a couple tweaks to the testing, and where the data is fetched from to make it easier to adapt to other services

File: src/app/components/LinkData/__snapshots__/index.test.jsx.snap (hunk 2 above)
Comment (twinlensreflex) [reply src/app/components/LinkData/__snapshots__/index.test.jsx.snap]: Thanks! I think I've fixed it now 👍

File: src/app/components/LinkData/index.jsx [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    logo,
+  };
+
+  const author = {
+    '@type': newsMediaOrg,
+    name: service,
+    logo,
+    noBylinesPolicy: 'http://www.bbc.com/news/help-41670342#authorexpertise',
```
Comment (pjlee11) [inline src/app/components/LinkData/index.jsx]: Same for this as Drew mentioned above

File: src/app/components/LinkData/index.jsx [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  lastUpdated,
+  optimoId,
+  service,
+  publishingPrinciples,
+  noBylinesPolicy,
+}) => {
+  const imgObject = 'ImageObject';
+  const newsMediaOrg = 'NewsMediaOrganization';
+  const url =
+    'https://www.bbc.com/news/special/2015/newsspec_10857/bbc_news_logo.png?cb=1';
```
Comment (dr3) [inline src/app/components/LinkData/index.jsx]: Can this be provided by the context too please? It will change per service

File: src/app/lib/config/services/persian.js [hunk 6]
```diff
@@ -14,6 +14,8 @@ const persian = {
   themeColor: `${C_POSTBOX}`,
   twitterCreator: '@bbcpersian',
   twitterSite: '@bbcpersian',
+  noBylinesPolicy: 'http://www.bbc.com/news/help-41670342#authorexpertise',
+  publishingPrinciples: 'http://www.bbc.com/news/help-41670342',
```
Comment (ChrisBAshton) [inline src/app/lib/config/services/persian.js]: Can we change all of these to HTTPS please?

https://github.com/bbc/simorgh/pull/1391/files/9e80e72b7d83f34ac36cfcbaed783a2568fb3216..0ecd0e6787835ae963a6603f07c059e7c611d119

**Review by dr3 — CHANGES_REQUESTED** [review]

File: .env [hunk 7]
```diff
@@ -1,4 +1,6 @@
-# On running the 'build' commands this file is replaced by the files in `env-config/`
-# To ignore and keep this file we have a command in `postshrinkwrap` which assumes this file is not changing
-# The command is `git update-index --assume-unchanged .env` 
-# Source: https://wildlyinaccurate.com/git-ignore-changes-in-already-tracked-files/
+SIMORGH_BASE_URL=http://localhost:7080
```
Comment (sareh) [inline .env]: The changes to `.env` should not be committed.

File: src/app/components/LinkData/index.jsx [hunk 8]
```diff
@@ -0,0 +1,80 @@
+import React from 'react';
+import Helmet from 'react-helmet';
+import { string } from 'prop-types';
+
+const LinkData = ({
```
Comment (sareh) [inline src/app/components/LinkData/index.jsx]: Just noticed that the component is called `LinkData` - can you please update it to be `LinkedData` (in the filenames as well as component names) This is since Linked Data is the name of the format we're using here. https://json-ld.org/

**Review by sareh — CHANGES_REQUESTED** [review]
Looks great so far. Just a couple of comments.

File: src/app/components/LinkedData/index.test.jsx [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+import { shouldShallowMatchSnapshot } from '../../helpers/tests/testHelpers';
+import LinkData from '.';
+
+describe('LinkData', () => {
+  const props = {
+    isAmp: false,
+    lang: 'en-GB',
+    seoHeadline: 'Royal wedding 2018: Bouquet laid on tomb of unknown warrior',
+    type: 'article',
+    optimoId: 'urn:bbc:optimo:asset:c9rpqy7pmypo',
```
Comment (dr3) [inline src/app/components/LinkedData/index.test.jsx]:
```suggestion
    optimoId: 'c9rpqy7pmypo',
```
This has the snapshot `https://www.bbc.com/news/articles/urn:bbc:optimo:asset:c9rpqy7pmypo`

File: src/app/containers/Metadata/index.jsx [hunk 10]
```diff
@@ -39,33 +40,50 @@ const MetadataContainer = ({ metadata, promo }) => {
...
+                  lang={metadata.passport.language}
+                  type={metadata.type}
+                  seoHeadline={promo.headlines.seoHeadline}
+                  firstPublished={timeFirstPublished}
+                  lastUpdated={timeLastPublished}
+                  optimoId={id}
+                  service={metadata.createdBy}
+                  publishingPrinciples={publishingPrinciples}
+                  noBylinesPolicy={noBylinesPolicy}
+                  url={defaultImage}
```
Comment (dr3) [inline src/app/containers/Metadata/index.jsx]: Why is this just `url`? Do you mean `logoUrl`?

**Review by dr3 — CHANGES_REQUESTED** [review]
Really close, just 2 small tweaks :)

File: src/app/containers/Metadata/index.jsx (hunk 10 above)
Comment (ChrisBAshton) [reply src/app/containers/Metadata/index.jsx]: Because inside `LinkedData.jsx` we construct the JSON using the `linkMetadata` object, and the JSON+LD format requires a `url` property for `logo` and `image`.

I thought we were constructing the object with `{...props}` (or equivalent) and thus needed the same name. But looking at it again, I can refactor it out 👍

File: src/app/components/LinkedData/index.test.jsx (hunk 9 above)
Comment (ChrisBAshton) [reply src/app/components/LinkedData/index.test.jsx]: Also affects the snapshot, so applied the fix in a separate commit:

https://github.com/bbc/simorgh/pull/1391/commits/6e596ef0d81b46add48e7ef650b97c74c3852bde

File: src/app/containers/Metadata/index.jsx (hunk 10 above)
Comment (ChrisBAshton) [reply src/app/containers/Metadata/index.jsx]: Resolved by https://github.com/bbc/simorgh/pull/1391/commits/7d65aa5cda2271ea887c9235970839c2dc5809e5

**Review by dr3 — APPROVED** [review]

**Review by pjlee11 — APPROVED** [review]
Great work 😎

**Comment by jamesbrumpton** [comment]
👍
